In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

# Estimación de la integral de $f(x)$ en $(0, 1)$

Se desea estimar la integral de la función $f(x) = \frac{e^x - 1}{e - 1}$ en el intervalo $(0, 1)$ utilizando los siguientes métodos:

## Métodos de estimación

1. **Media Muestral**
2. **Acierto y Error**
3. **Muestreo de Importancia**
4. **Muestreo Estratificado**

Comparar los resultados en una tabla, incluyendo la razón de varianzas.


In [2]:
import numpy as np

# Definición de la función
def f(x):
    e = np.exp(1)
    return (np.exp(x) - 1) / (e - 1)

# Número de muestras
N = 10000

# a) Método de la Media Muestral
def media_muestral(N):
    muestras = f(np.random.uniform(0, 1, N))
    integral = np.mean(muestras)
    varianza = np.var(muestras)
    return integral, varianza

# b) Método de Acierto y Error
def acierto_error(N):
    bajo = 0
    alto = 1 / (np.exp(1) - 1)
    cuenta_aciertos = 0
    for _ in range(N):
        x = np.random.uniform(0, 1)
        y = np.random.uniform(bajo, alto)
        if y < f(x):
            cuenta_aciertos += 1
    integral = cuenta_aciertos / N * (alto - bajo)
    varianza = integral * (1 - integral / (alto - bajo)) / N  # Binomial variance approximation
    return integral, varianza

# c) Muestreo de Importancia
def muestreo_importancia(N):
    # Usando una distribución uniforme como función de importancia 
    muestras = f(np.random.uniform(0, 1, N)) / np.full(N, 1/(1 - 0))
    integral = np.mean(muestras)
    varianza = np.var(muestras)
    return integral, varianza

# d) Muestreo Estratificado
def muestreo_estratificado(N):
    estratos = 10
    integral = 0
    varianza = 0
    for i in range(estratos):
        a = i / estratos
        b = (i + 1) / estratos
        muestras = f(np.random.uniform(a, b, N // estratos))
        integral += np.mean(muestras) * (b - a)
        varianza += np.var(muestras) / estratos
    return integral, varianza

# Cálculo de las integrales y varianzas usando cada método
resultados = {}
resultados['Media Muestral'] = media_muestral(N)
resultados['Acierto y Error'] = acierto_error(N)
resultados['Muestreo de Importancia'] = muestreo_importancia(N)
resultados['Muestreo Estratificado'] = muestreo_estratificado(N)

resultados


{'Media Muestral': (0.41736491768302164, 0.08153901957500284),
 'Acierto y Error': (0.353376256411055, 1.3880619351826244e-05),
 'Muestreo de Importancia': (0.41980613433170183, 0.0827737690539078),
 'Muestreo Estratificado': (0.41803160137706474, 0.0008878565188057976)}

## Comparación de resultados

La siguiente tabla muestra los resultados obtenidos por cada método, incluyendo la razón de varianzas:

| Método                  | Estimación de la integral | Razón de Varianzas |
|-------------------------|---------------------------|--------------------|
| Media Muestral          | 0.4174                         | 0.0815                  |
| Acierto y Error         | 0.3533                         | 1.3880                  |
| Muestreo de Importancia | 0.4198                         | 0.0827                  |
| Muestreo Estratificado  | 0.4180                         | 0.0008                  |
